drivers/iommu/riscv/io_pgtable.c

```diff
@@ -20,55 +20,108 @@
 	container_of(container_of((x), struct io_pgtable, ops), \
 			     struct riscv_iommu_domain, pgtbl)
 
+#define io_pgtable_to_domain(x) \
+	container_of((x), struct riscv_iommu_domain, pgtbl)
+
+static inline size_t get_page_size(size_t size)
+{
+       if (size >= IOMMU_PAGE_SIZE_512G)
+               return IOMMU_PAGE_SIZE_512G;
+
+       if (size >= IOMMU_PAGE_SIZE_1G)
+               return IOMMU_PAGE_SIZE_1G;
+
+       if (size >= IOMMU_PAGE_SIZE_2M)
+               return IOMMU_PAGE_SIZE_2M;
+
+       return IOMMU_PAGE_SIZE_4K;
+}
+
+static void riscv_iommu_pt_walk_free(pte_t *ptp, unsigned shift, bool root)
+{
+	pte_t *pte, *pt_base;
+	int i;
+
+	if (shift == PAGE_SHIFT)
+		return;
+
+	if (root)
+		pt_base = ptp;
+	else
+		pt_base = (pte_t *)pfn_to_virt(__page_val_to_pfn(pte_val(*ptp)));
+
+	for (i = 0; i < PTRS_PER_PTE; i++) {
+		pte = pt_base + i;
+		// Recursively free all sub page table pages
+		if (pte_present(*pte) && !pte_leaf(*pte))
+			riscv_iommu_pt_walk_free(pte, shift - 9, false);
+	}
+
+	// Now free the current page table page
+	if (!root && pte_present(*pt_base))
+		free_page((unsigned long)pt_base);
+}
+
 static void riscv_iommu_free_pgtable(struct io_pgtable *iop)
 {
-	printk("io-pgtable for riscv released\n");
+	struct riscv_iommu_domain *domain = io_pgtable_to_domain(iop);
+	riscv_iommu_pt_walk_free((pte_t *)domain->pgd_root, PGDIR_SHIFT, true);
 }
 
-static pte_t *riscv_iommu_pgd_walk(struct riscv_iommu_domain *domain,
+static pte_t *riscv_iommu_pt_walk_alloc(pte_t *ptp,
 				   unsigned long iova,
-				   unsigned long (*pd_alloc)(gfp_t), gfp_t gfp)
+				   unsigned shift, bool root, size_t pgsize,
+				   unsigned long (*pd_alloc)(gfp_t), gfp_t gfp,
+				   bool *updated)
 {
-	/* TODO: merge dev/iopgtable */
-	pgd_t *pgd;
-	p4d_t *p4d;
-	pud_t *pud;
-	pmd_t *pmd;
+	pte_t *pte;
```

Comment:
You can treat all PTEs as PMDs and use pmd_xxx accessors (here and in riscv_iommu_pt_walk_free()), including pmd_leaf(). There will be no need for extra pte_leaf(), what should btw always return true if its pte_t.   For pte_xxx use pmd_pte() to remap pointer.

Reply:
Ok I'll rework accordingly.

Reply:
Done!